### PR TITLE
73 connected sites

### PR DESCRIPTION
**Pull request by zpoken**

### CODE REVIEW

File: apps/extension/src/state/messages.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+          status,
+        };
+
+        const updatedMsgIndex = messages.findIndex(msg => msg.id === id);
+
+        // Replace item at index
+        messages[updatedMsgIndex] = updatedMsg;
+
+        set(state => {
+          state.messages.messages = messages;
```
Comment (grod220) [inline apps/extension/src/state/messages.ts:33]: Having `.messages.messages` is not good field naming

File: apps/extension/src/routes/popup/active-message/connect.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      </div>
+      <div className='flex gap-5'>
+        <Button
+          variant='secondary'
+          className='w-[50%]'
+          size='lg'
+          onClick={() =>
+            void (async () => {
+              await changeStatusMessage(message.id, MessageStatus.REJECTED);
+              await services.updateBadge();
```
Comment (grod220) [inline apps/extension/src/routes/popup/active-message/connect.tsx:34]: What does this do? Please provide a screenshot for updateBadge(). Does that show somewhere?

File: packages/router/src/std/routes/connect.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  ): SwMessageHandler<ConnectMessage> =>
+  async () => {
+    const connectedSites = await localExtStorage.get('connectedSites');
+    const messages = await localExtStorage.get('messages');
+
+    // check is pending message exist with same origin
+    if (messages.find(msg => msg.status === MessageStatus.PENDING && msg.origin === sender.origin))
+      return;
+
+    // create message when connectedSites doesn`t include orrigin
```
Comment (grod220) [inline packages/router/src/std/routes/connect.ts:24]:
```diff
- orrigin
+ origin
```

File: packages/router/src/std/routes/connect.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if (!connectedSites.length || !connectedSites.find(origin => origin === sender.origin)) {
+      // check is origin is not approved early
```
Comment (grod220) [inline packages/router/src/std/routes/connect.ts:26]: What does this sentence mean? Needs revised.

File: packages/router/src/std/routes/connect.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  (
+    services: ServicesInterface,
+    sender: chrome.runtime.MessageSender,
+  ): SwMessageHandler<ConnectMessage> =>
+  async () => {
+    const connectedSites = await localExtStorage.get('connectedSites');
+    const messages = await localExtStorage.get('messages');
+
+    // check is pending message exist with same origin
+    if (messages.find(msg => msg.status === MessageStatus.PENDING && msg.origin === sender.origin))
```
Comment (grod220) [inline packages/router/src/std/routes/connect.ts:21]: What is a message and what does it mean for a message to be pending?

File: apps/extension/src/routes/popup/home/index.tsx [hunk 6]
```diff
@@ -37,8 +43,8 @@ export const PopupIndex = () => {
   const { next, previous } = useStore(accountsSelector);
 
   return (
-    <div className='relative flex h-full flex-col items-stretch justify-start bg-left-bottom px-[30px]'>
-      <div className='absolute bottom-[50px] left-[-10px] -z-10 h-[715px] w-[900px] overflow-hidden bg-logo opacity-10' />
+    <div className='relative flex h-full min-h-screen flex-col items-stretch justify-start bg-left-bottom px-[30px]'>
+      <div className='absolute bottom-[50px] left-[-10px] -z-10 h-[600px] w-[400px] overflow-hidden bg-logo opacity-10' />
```
[the remark this answers is not in the document]
Comment (grod220) [reply apps/extension/src/routes/popup/home/index.tsx:47]: Why does it need to be changed?

File: apps/extension/src/routes/popup/home/index.tsx [hunk 7]
```diff
@@ -27,6 +28,11 @@ export const popupIndexLoader = async (): Promise<Response | PopupLoaderData> =>
 
   if (!password) return redirect(PopupPath.LOGIN);
 
+  const messages = await localExtStorage.get('messages');
+
+  if (messages.find(msg => msg.status === MessageStatus.PENDING))
+    return redirect(PopupPath.ACTIVE_MESSAGE);
```
Comment (grod220) [inline apps/extension/src/routes/popup/home/index.tsx:34]: Why are we re-using the index route of popup.html? Shouldn't connected sites just be its own route?

Comment (grod220) [reply apps/extension/src/routes/popup/home/index.tsx:34]: When you click on the `Connect` button after you are already connected, it creates a new normal popup window. Feels incorrect.

**Review by grod220 — COMMENTED** [review]
Few additional things:
- The whole messages slice and pattern feels confusing to me. I think we should think of a better way to do this. I think we can handle this via url routing versus having to keep messages in storage and read them one second later. It's not clear to me why we are using the popup index route for this feature. Think we need to go back to the drawing board and simplify this architecture. 
- I noticed that `git merge` is used quite often. This is difficult when reviewing as the history becomes muddled. I request that you only use `git rebase` from now on for these. Will make the history much easier to understand.
- At the moment, non-connected sites can still make requests to the chrome extension. You may require doing some work here: https://github.com/penumbra-zone/web/blob/main/packages/transport/src/proxy.ts (or perhaps somewhere else) in order to narrow only connected sites to make requests.

File: packages/services/src/index.ts [hunk 8]
```diff
@@ -92,4 +93,32 @@ export class Services implements ServicesInterface {
...
+  }
+
+  async updateBadge() {
+    const messagesLength = (await localExtStorage.get('messages')).filter(
+      msg => msg.status === MessageStatus.PENDING,
+    ).length;
+
+    await chrome.action.setBadgeText({ text: messagesLength ? String(messagesLength) : '' });
+    await chrome.action.setBadgeBackgroundColor({ color: '#629994' });
+    await chrome.action.setBadgeTextColor({ color: '#fff' });
```
[the remark this answers is not in the document]
Comment (zpoken) [reply packages/services/src/index.ts:122]: Знімок екрана 2023-10-23 о 18 59 21

File: apps/extension/src/routes/popup/home/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (zpoken) [reply apps/extension/src/routes/popup/home/index.tsx:47]: https://github.com/penumbra-zone/web/blob/main/apps/extension/src/routes/popup/home/index.tsx#L41 it's old code! you added `bottom-[50px] left-[-10px]` and `h-[715px] w-[900px]` is not popup sizes!

File: packages/router/src/std/routes/connect.ts (hunk 5 above)
Comment (zpoken) [reply packages/router/src/std/routes/connect.ts:21]: when user click connect, we have message with pending status (need to show active message page), after user can Approve or Reject message

File: packages/services/src/index.ts [hunk 9]
```diff
@@ -92,4 +93,32 @@ export class Services implements ServicesInterface {
     this.walletServicesPromise = undefined;
     await this.initialize();
   }
+
+  async openWindow() {
+    const { top, left, width } = await chrome.windows.getLastFocused();
+
+    await chrome.windows.create({
+      url: 'popup.html',
```
[the remark this answers is not in the document]
Comment (zpoken) [reply packages/services/src/index.ts:101]: why can it be a problem that we use popup.html?

[the remark this answers is not in the document]
Comment (zpoken) [reply packages/services/src/index.ts:101]: I can create notification.html for this

**Comment by zpoken** [comment]
should i rename messages to notifications? maybe it will be clearer

File: apps/extension/src/routes/popup/home/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (grod220) [reply apps/extension/src/routes/popup/home/index.tsx:47]: No it's not quite right. See the current extension design on main and match that.

File: packages/services/src/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (grod220) [reply packages/services/src/index.ts:101]: I suppose the question is why we are using the index route on popup.html and not a specific route for connecting sites?

**Comment by grod220** [comment]
> should i rename messages to notifications? maybe it will be clearer

I don't think that slice is necessary to have. Why do we need notifications at all when we can just have a `connecting sites` route that we open a window for. I think it's creating unnecessary state.

**Comment by Vadim121197** [comment]
What about txs in the future?we should open popup too

**Comment by grod220** [comment]
> What about txs in the future?we should open popup too

Sure, that's fine. But why should that route be the index? Why would that not be its own route?

**Comment by Vadim121197** [comment]
What do you mean "own route"?

**Comment by grod220** [comment]
These are routes: https://github.com/penumbra-zone/web/blob/main/apps/extension/src/routes/popup/paths.ts

**Comment by zpoken** [comment]
https://github.com/penumbra-zone/web/blob/73-connected-sites/apps/extension/src/routes/popup/paths.ts#L15
there is own route than

**Comment by zpoken** [comment]
you mean active-message/connect as example&?

**Comment by grod220** [comment]
I'm trying to understand what is the purpose for messages/notifications. Why is this necessary? For instance, why can't the service worker open `popup.html/connect-site?origin=app.testnet.penumbra.zone`?

File: apps/webapp/app/header/header.tsx [hunk 10]
```diff
@@ -27,6 +31,7 @@ export const Header = () => {
         <Notifications />
         <Network />
         <WalletId />
+        <ConnectButton />
```
[the remark this answers is not in the document]
Comment (zpoken) [reply apps/webapp/app/header/header.tsx:34]: I haven't done dapp yet, I was planning to do it after this

**Comment by zpoken** [comment]
i see! I think this is necessary so that we can return some information to the user after they have confirmed or denied the ping request

**Comment by zpoken** [comment]
the user can lose focus on the popup by simply clicking on the browser (the popup goes to the background of the screen) and then he can open the extension and he will already have a page for connection or transaction open! also, the transaction can take a long time to complete and then the user closing the popup will lose information on it

**Comment by zpoken** [comment]
Знімок екрана 2023-10-23 о 22 39 11

lose focus example

**Comment by grod220** [comment]
The transaction use case _may_ be valid, but I don't think connected sites is the same. Even if they lose focus, it's not a big deal. Also, I don't see why the the popup needs to return information to the user that they accepted or rejected the request. For connected sites, I don't see the message/notification use case.

**Comment by Vadim121197** [comment]
metamask return information is user reject tx or result of tx! how does we know if user approve connect?

**Comment by grod220** [comment]
This PR isn't about transactions. So let's stick to the connected sites use case. If the user approves the connected site, it should just save that to storage and close the window no?

**Comment by grod220** [comment]
I do not understand your last message. Please put more detail on the specific issue of having: connect button --> connect popup --> accept --> close.

**Comment by Vadim121197** [comment]
When user approve connection, how can I do sync , balance and other request? We should have some state when user approve connection! Metamask return public key after approve connection

**Comment by Vadim121197** [comment]
As example https://docs.metamask.io/wallet/how-to/connect/set-up-sdk/javascript/react/

**Comment by Vadim121197** [comment]
We should do request when user approve connection

**Comment by grod220** [comment]
I'm not understand what you mean. When you connect to a site, you don't trigger a sync. The extension is already syncing in the background. When you connect to a site, you don't request balances-----the web app is responsible for requesting balances, not the popup.

You may need to draw a diagram because I don't understand what you are meaning with the whole technical design of this.

**Comment by Vadim121197** [comment]
What I can see without connection from Dapp?

**Comment by Vadim121197** [comment]
currently there is no logic if the extension is not installed!

**Comment by turbocrime** [comment]
will be part of #475

**Pull request closed without merging**